src/gpt.py

```diff
@@ -0,0 +1,312 @@
+import asyncio
+import datetime
+import json
+import random
+import re
+from typing import Dict, List, Optional, Tuple
+
+import openai
+import tiktoken
+from nextcord import Member, Message
+
+import config
+
+
+# Initialize globals
+def initialize(client):
+    global gpt_lock
+    global context_data
+    global encoding
+    global self_user
+    global banned_word_pattern
+    global word_trigger_pattern
+    global user_trigger_pattern
+    global user_info_map
+
+    openai.api_key = config.openai_api_key
+    # Global lock for message response
+    gpt_lock = asyncio.Lock()
+    # Load manual hidden data
+    try:
+        with open(config.gpt_context_file) as f:
+            context_data = json.load(f)
+    except (FileNotFoundError, json.decoder.JSONDecodeError) as e:
+        print(
+            f"ERROR: Issue loading {config.gpt_context_file}. GPT capabilities will be disabled.\n{e}"
+        )
+        # Signal context data was not loaded correctly
+        context_data = None
+        return
+    # Preload tokenizer
+    encoding = tiktoken.encoding_for_model(config.OPENAI_MODEL)
+    # Store bot user
+    self_user = client.user
+    # Cache regex for banned words
+    banned_word_pattern = re.compile(
+        r"\b{}\b".format("|".join(context_data["banned_phrases"]))
+    )
+
+    # Cache regex for word triggers
+    word_trigger_pattern = re.compile(
+        r"\b({})\b".format("|".join(context_data["word_triggers"].keys()))
+    )
+
+    # Cache regex for user info triggers
+    user_trigger_pattern = re.compile(
+        r"\b({})\b".format(
+            "|".join(
+                [
+                    user["name"].lower()
+                    for user in context_data["user_info"].values()
+                    if "info" in user
+                ]
+            )
+        )
+    )
+    # Store map for user info triggers
+    user_info_map = {
```

Comment:
¯\\_(ツ)_/¯

my reasoning is that even if it may be slightly less efficient for our small size, it's still currently dwarfed by API response time, definitely still /pretty/ fast, and it will scale much better so we don't have to re-test if we add more things.

Reply:
Oops, I meant to address that question myself before submitting the review haha.

It looked like a slightly odd data structure on first glance. But looking again with fresh eyes, it looks fine.